base/src/Data/Macaw/Memory.hs

```diff
@@ -1,8 +1,46 @@
 {-|
 Copyright   : (c) Galois Inc, 2015-2018
-Maintainer  : jhendrix@galois.com
-
-Declares 'Memory', a type for representing segmented memory with permissions.
+Maintainer  : Ryan Scott <rscott@galois.com>, Langston Barrett <langston@galois.com>
+
+Declares 'Memory', a type for representing pre-loader memory with permissions.
+This datatype provides an abstraction that is intended to support different
+architectures and executable and object file formats. Crucially, 'Memory' is
+capable of representing relocatable (i.e., position-independent) code and data.
+
+A 'Memory' defines a collection of abstract /regions/ of memory. A region is
+identified by a 'RegionIndex', and represents some an address that would be
+chosen at runtime by the loader (e.g., the virtual address of an ELF segment
+containing position-independent code). Thus, in this module, an address
+('MemAddr') consists of a pair of a 'RegionIndex' and an offset into that
+region. A 'MemAddr' with a 'RegionIndex' of 0 represents an /absolute/ address.
+
+'Memory' is essentially a collection of segments. A segment ('MemSegment')
+is a continguous sequence of bytes that will be loaded into runtime memory.
+Segments do not necessarily have a known runtime address. Instead, they use some
+'RegionIndex' as a "base" address, and are located at some fixed offset from
+that base. Multiple segments can have the same 'RegionIndex' as their base; this
+indicates that they will have a fixed offset relative to one another at runtime.
+This notion of segment is similar to an ELF segment. It is unrelated to the x86
+notion of memory segmentation.
+
+= Addresses and related types
+
+As described above, an address ('MemAddr') consists of a base ('RegionIndex')
+and an offset. This section describes a few types that are adjacent to this one,
+along with their intended use-cases.
+
+A @'MemWord' w@ is a @w@-bit machine word. This may be treated as an absolute
+address when @w@ is the width of a pointer ('absoluteAddr').
+
+A 'MemSegmentOff' is notionally a pair @('MemSegment', offset)@, where @offset@
+is an offset into the 'MemSegment'. The 'MemSegmentOff's produced by this module
+are guaranteed to be valid, making it possible to look up the contents of the
+memory they point to.
+
+Each of the above types is parameterized by the number of bits in an address.
+Each type has an alias prefixed with @Arch@ that is instead parameterized by
```

Comment:
Strictly speaking, I don't know if this is true (e.g., I don't think there's an `Arch` counterpart to `MemWord`). Perhaps we should say that _most_ types have aliases prefixed with `Arch` instead.